### ISSUE TITLE
Deserializing mixed content

### ISSUE THREAD

**Issue by Goz3rr** [ExtendedXmlSerializer/home#443]
Hey,

I'm trying to convert some XDocument code to deserialize some mixed XML content to use ExtendedXmlSerializer to greatly simplify the code. I started with the plain old XmlSerializer, but quickly ran into issues where I had to serialize an attribute that could either be a decimal or hexadecimal string which was easily solved with a custom converter but now I've run into the next issue which I can't quite figure out how to solve.

The relevant section of XML looks as follows:

```xml
<StringFormat>Example <Ref field="Foo" /> bit of mixed <Ref field="Bar" /> xml content <Ref field="Baz" />ms</StringFormat>
```

I'm trying to get this into some data structures along these lines:
```cs
    public class MessageStringRef
    {
        [XmlAttribute("field")]
        public string Reference { get; set; }
    }

    public class MessageStringFormat
    {
        [XmlText(typeof(string))]
        [XmlElement("Ref", typeof(MessageStringRef))]
        public List<object> Parts { get; set; }
    }
```

So far everything I've tried with attributes results in `Unknown/invalid member encountered: 'field'`. I have no idea if ExtendedXmlSerializer even supports mixed content, so I tried registering a custom serializer for the StringFormat element. However, if I try to access parameter.Content() inside the serializer all I get is the text up until the first XML element, in the case of the example that's `"Example "` which makes it impossible to do any custom parsing myself. (In the worst case, if I could get the entire body of the `<StringFormat>` node I can just fall back to the old XDocument parsing)

**Comment by Mike-E-angelo** [comment]
Hi @Goz3rr thank you for taking the time to try out ExtendedXmlSerializer and for writing in with your question.  Unfortunately, you won't like the answer. 😞  `XmlTextAttribute` and inner content is not supported very well if at all by ExtendedXmlSerializer.  In fact, what support we do have for `XmlTextAttribute` should be removed at a later version as it really doesn't do its job well with what has already been implemented.

While we do support some legacy/classic XmlSerializer functionality and features, the focus/strength of ExtendedXmlSerializer is best used around object graphs, and those that are serialized and then deserialized directly by ExtendedXmlSerializer, meaning no external/interop with other serializers are used.

FWIW, we do have some more information/context around this in our wiki here:
https://github.com/ExtendedXmlSerializer/home/wiki/FAQs#systemxmlserializer-vs-extendedxmlserializer

As for mixed content/inner content within an element, you have indeed reached a current limitation as really the only "content" within an element is used for constructing the current item in the object graph.  That is, there is no native notion of "give me all the text and inner elements of the current XML node."  However, if you are ambitious/committed enough, you can get a hold of the native `System.Xml.XmlReader` within the `IFormatReader` of your custom serializer by calling the `Get` method on it and casting it to `System.Xml.XmlReader`.  From there, you can do further work with it to perhaps get what you need.  Example of this can be found here:

https://github.com/ExtendedXmlSerializer/home/blob/4f84d015ce76690640321fda9ed45cd4d461c14d/test/ExtendedXmlSerializer.Tests.ReportedIssues/Issue346Tests_Custom.cs#L79-L83

Please do let me know if you have any further questions around this and I will do my best to assist you.

**Comment by Goz3rr** [comment]
Thanks for the quick reply,

I've indeed managed to do what I want by getting the XmlReader, after which the reader is on the next node in the tree. The strange thing however is that when my serializer returns, it seems the library just completely bails on parsing the rest of the document. No exceptions are thrown (`.WithUnknownContent().Throw()` is added), it just stops after the serializer and returns an incomplete object.

**Comment by Mike-E-angelo** [comment]
Yeah, we're sort of in uncharted territory here, @Goz3rr.  Or rather, off the beaten path.  You might have to ensure the reader is on an element node to properly continue.  There is the `IFormatReader.Set` method you could try.

If not, and you can provide some further code on what you're working with, I could take a peek and see if I can further assist.

**Comment by Goz3rr** [comment]
I tried to make a small repro but can't seem to trigger the same issue, I'll get back to you on Monday when I can look at my code that doesn't work

**Comment by Goz3rr** [comment]
Here's a small program that reproduces my issue, when setting a breakpoint inside the `StringFormatSerializer.Get` method it is only called once and an incomplete object is returned by `serializer.Deserialize` right after.

```cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Xml;
using System.Xml.Serialization;

using ExtendedXmlSerializer;
using ExtendedXmlSerializer.Configuration;
using ExtendedXmlSerializer.ContentModel;
using ExtendedXmlSerializer.ContentModel.Format;

namespace ConsoleApp2
{
    public class TestRoot
    {
        [XmlArray]
        public List<TestItem> TestItems { get; set; }
    }

    public class TestItem
    {
        [XmlAttribute("id")]
        public string Id { get; set; }
        [XmlAttribute("messageGroup")]
        public string MessageGroup { get; set; }
        [XmlAttribute("messageNumber")]
        public string MessageNumber { get; set; }

        [XmlElement]
        public TestSubItem Notification { get; set; }
        [XmlElement]
        public TestSubItem Command { get; set; }
        [XmlElement]
        public TestSubItem Response { get; set; }
    }

    public class TestSubItem
    {
        [XmlElement]
        public MessageStringFormat StringFormat { get; set; }
    }

    public class MessageStringFormat
    {
        public string Body { get; set; }
    }

    public class StringFormatSerializer : ISerializer<MessageStringFormat>
    {
        public MessageStringFormat Get(IFormatReader parameter)
        {
            var reader = parameter.Get().AsValid<XmlReader>();
            var content = reader.ReadOuterXml();

            return new MessageStringFormat()
            {
                Body = content
            };
        }

        public void Write(IFormatWriter writer, MessageStringFormat instance)
        {
            throw new NotImplementedException();
        }
    }

    class Program
    {
        private const string TestXml = @"<TestRoot>
    <TestItems>
        <TestItem id=""TestItem1"" messageGroup=""Test"" messageNumber=""0x01"">
            <Notification>
                <StringFormat>
                    Foo <Ref field=""Foo"" />
                </StringFormat>
            </Notification>
        </TestItem>

        <TestItem id=""TestItem2"" messageGroup=""Test"" messageNumber=""0x02"">
            <Command>
                <StringFormat>
                    Foo <Ref field=""Foo"" />, Bar <Ref field=""Bar"" />, Baz <Ref field=""Baz"" />ms
                </StringFormat>
            </Command>

            <Response>
                <StringFormat>
                    Foo <Ref field=""Foo"" />, Bar <Ref field=""Bar"" />
                </StringFormat>
            </Response>
        </TestItem>

        <TestItem id=""TestItem3"" messageGroup=""Test"" messageNumber=""0x03"">
            <Notification />
        </TestItem>

        <TestItem id=""TestItem4"" messageGroup=""Test"" messageNumber=""0x04"">
            <Notification />
        </TestItem>
    </TestItems>
</TestRoot>";

        static void Main(string[] args)
        {
            var settings = new XmlReaderSettings
            {
                IgnoreWhitespace = true
            };
            var reader = XmlReader.Create(new StringReader(TestXml), settings);

            var serializer = new ConfigurationContainer()
                .EnableImplicitTyping(typeof(TestRoot))
                .WithUnknownContent().Throw()
                .Type<MessageStringFormat>().Register().Serializer().Using(new StringFormatSerializer())
                .Create();

            var result = (TestRoot)serializer.Deserialize(reader);
        }
    }
}
```

While trying to reproduce it I found out the key here seems to be the `XmlReaderSettings` having `IgnoreWhitespace` set to true. If I omit this the program suddenly seems to work as expected.

**Comment by Mike-E-angelo** [comment]
OK great, thank you for the code @Goz3rr, I'm digging into this now and at first glance, I share your baffle-affinity. 😆  To be sure, the reader is the one area that has a lot of sore spots, so it's a little fragile here.  In this case, we're dealing with nested properties and it seems the state isn't persisted correctly after calling `XmlReader.ReadOuterXml`.  As written only one `TestItem` gets created.  With some tweaking, I was able to get two `TestItem`s created, but there should be four and that is a problem.

I will continue to poke to see what is going on here.  It may take some time here in the next day or so before I am able to do so, however.  I will update here with any findings.

**Comment by Mike-E-angelo** [comment]
Alright @Goz3rr I've spent some time looking into this and have found the explanation for this.  Unfortunately I do not see an easy fix for it.

The issue is around how the reader keeps track of state during reading, more specifically the `System.Xml.XmlReader.Depth` property:

https://github.com/ExtendedXmlSerializer/home/blob/acab951fe5dc1e5082ee9aadcae2eb5ee7155b18/src/ExtendedXmlSerializer/ExtensionModel/Xml/XmlElements.cs#L14

When `XmlReaderSettings.IgnoreWhitespace` is `true`, the call to `System.Xml.XmlReader.ReadOuterXml` decreases the depth by one, whereas it doesn't when `XmlReaderSettings.IgnoreWhitespace` is `false`.  I am unclear why this is the case.  However, when the depth is decreased during this time, it invalidates the state when returning to the calling reader and it ceases to stop the enumeration of members.  This explains why it doesn't throw when `WithUnknownContent().Throw()` is specified, as it isn't encountering unknown content, but rather simply stops enumeration of members as a condition is no longer met.

As I alluded to earlier, the reader isn't perfect, and it is long overdue for a rewrite.  I'm actually amazed that it works as well as it does. 😁  That stated, I am not sure of a good way to work around this except to use the workaround that you have already identified, which is in itself also a less than ideal solution.

Let me know if you have any further comments/questions/suggestions and I will do my best to assist you with them. 👍
